odoo/addons/base/models/ir_translation.py

```diff
@@ -923,3 +923,40 @@ def get_web_translations_hash(self, mods, lang):
             'multi_lang': len(self.env['res.lang'].sudo().get_installed()) > 1,
         }
         return hashlib.sha1(json.dumps(translation_cache, sort_keys=True).encode()).hexdigest()
+
+    @api.model
+    def _guess_language_of_terms(self, terms, types=None):
+        """ Guess the language of the terms based on the translations stored in the database.
+        Note that the search relies on some non indexed fields.
+        :param tuple(str) terms: terms used in the application for which we want to guess the language
+        :param set types: if set, limit the search to the given types (increase performance)
+        :return tuple(str, float):
+            - best_candidate (str): the best language (code) candidate or None if none (ex.: fr_BE)
+            - best_candidate_term_ratio (float): ratio of term found in the database for the best_candidate language
+        """
+        params = dict(terms=tuple([t.lower().strip() for t in terms]), main_lang='en_US')
+        if types:
+            params['types'] = tuple(types)
+        restrict_to_types_clause = ' AND type IN %(types)s ' if types else ''
+        self._cr.execute(f"""
+        SELECT sub.lang, SUM(sub.term_occ) term_occ
+            FROM 
+            (
+                SELECT lang, COUNT(DISTINCT TRIM(LOWER("value"))) term_occ        
```

Comment:
can you move this into a statement ? ("WITH query")
https://www.postgresql.org/docs/current/queries-with.html

it will remove one indentation level and make the query more readable :)

Reply:
Thanks for the tip, it is indeed better that way. I have also slightly changed the query as I have noticed that there are translations for the main language as well. I do the count outside the sub query to avoid double counting terms for the main language.